qemu/tests/vsock_test.py

```diff
@@ -0,0 +1,170 @@
+    """
+    nc_vsock_cmd = "%s -l %s" % (nc_vsock_bin, port)
+    session.read_nonblocking(0, timeout=10)
+    logging.info("Listening to the vsock port from guest: %s" % nc_vsock_cmd)
+    session.sendline(nc_vsock_cmd)
+
+
+def check_received_data(test, session, pattern):
+    """
+    Check if session received expected data
+
+    :param test: QEMU test object
+    :param session: Session object to be checked
+    :param pattern: Expected pattern of session output
+    :return: Error msg if not receive expected content, None if received
+    """
+    try:
+        session.read_until_last_line_matches([pattern])
+    except aexpect.ExpectError as e:
+        if isinstance(e, aexpect.ExpectTimeoutError):
+            test.fail("Does not receive expected content: %s, output"
+                      " of session: %s" % (pattern, e.output))
+        else:
+            test.fail(str(e))
+
+
+def nc_vsock_connect(nc_vsock_bin, guest_cid, port):
+    """
+    Connect to vsock port from host, cmd: nc-vsock $guest_cid $port
+
+    :param nc_vsock_bin: path of binary nc-vsock
+    :param guest_cid: guest cid to connect
+    :param port: port to connect
+    :return: The vsock session from host side, being waiting for input
+    """
+    nc_vsock_cmd = "%s %s %s" % (nc_vsock_bin, guest_cid, port)
+    logging.info("Connect to the vsock port on host: %s" % nc_vsock_cmd)
+    return aexpect.Expect(
+        nc_vsock_cmd,
+        auto_close=False,
+        output_func=utils_misc.log_line,
+        output_params=("vsock_%s_%s" % (guest_cid, port),))
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Vsock basic function test
+
+    1. Boot guest with vhost-vsock-pci device
+    2. Download and compile nc-vsock on both guest and host
+    3. Start listening inside guest, nc-vsock -l $port
+    4. Connect guest CID from host, nc-vsock $guest_cid $port
+    5. Input character, e.g. 'Hello world'
+    6. Check if guest receive the content correctly
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def clean(tmp_file=None):
+        """ Clean the environment """
+        cmd_rm = "rm -rf %s*" % nc_vsock_bin
+        if tmp_file:
+            cmd_rm += "; rm -rf %s" % tmp_file
+        session.cmd_output_safe(cmd_rm)
+        process.system(cmd_rm, ignore_status=True)
+        if host_vsock_session.is_alive():
+            host_vsock_session.close()
+        session.close()
+
+    vm = env.get_vm(params["main_vm"])
+    tmp_file = "/tmp/vsock_file_%s" % utils_misc.generate_random_string(6)
+    session = vm.wait_for_login()
+    # TODO: Close selinux as temporary workaround for qemu bug 1656738
+    # should be removed when fixed
+    session.cmd_output("setenforce 0")
+    nc_vsock_bin = compile_nc_vsock(test, vm, session)
+    vm_ip_addr = vm.wait_for_get_address(0)
+    port = utils_misc.find_free_port(5000, 6000, vm_ip_addr)
```

Comment:
@vivianQizhu  As port number can be any port.  so we think you can replace port number with guest_cid. e.g ./nc-vsock -l 3 3.   If you don't want to change it.  It is also ok for me.  As porlarion case is 5000 port.

Reply:
@xiangchunfu I don't agree it can be any port, it should be a free port, we might hit error if the port is allocated by other process, and that will be an invalid scenario.

Reply:
@vivianQizhu It don't use system port.  so if the port is allocated, it also works.  e.g use 22 port.

Reply:
@xiangchunfu I see, will update per your suggestion, to use guest-cid as port. Thanks.